### ISSUE TITLE
Datasource Devservice: Skip init script, if testcontainers.reuse.enable=true

### ISSUE THREAD

**Issue by fsomme2s** [quarkusio/quarkus#36987]
### Describe the bug

When testcontainers.reuse.enable=true, the database container is reused between Quarkus hot reloads (a really awesome feature btw). But also the init script referenced in quarkus.datasource.devservices.init-script-path seems to be executed again.

I would count that as a bug - but that might be debatable. 

My reasoning is: The script is meant to initialize the database. So if I keep the database up, it does not need to be initialized again. If your init script does actions that are not repeatable, there will be errors on hot reload and your app has to be restarted (and sometimes the still running container needs to be killed manually).

Or maybe it should be configurable, if some kinds of init scripts should run on each hot reload and some don't?

Current workaround is to make your init scripts idempotent (which is often easy, but in case of creating users in postgres a simple "if not exists" is not available).

### Expected behavior

Don't execute the init script on quarkus hot reload, when testcontainers.reuse.enable=true is set.

### Actual behavior

Script gets executed again on every hot reload.

### How to Reproduce?

Steps to reproduce:

1. Create a Quarkus Project with SQL Datasource, for example with "quarkus-jdbc-postgresql" dependency
```
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-jdbc-postgresql</artifactId>
    </dependency>
```

2. Configure DevService with init script and enable container reuse

application.properties:

`quarkus.datasource.devservices.init-script-path=db/init/init-users.sql`

~/.testcontainers.properties:

`testcontainers.reuse.enable=true`

3. Do something non-idempotent in the init script, that will produce an error on second execution, for example:

src/main/resources/db/init/init-users.sql:
`create user app_user with password 'local';`

4. Start the app (everything should be fine for first start)
5. Change a line in your code and trigger something for the hot reload feature (for example a rest api call)

=> Now your init script will throw an error

**Comment by gsmet** [comment]
`testcontainers.reuse.enable=true` is totally outside of our control so I'm not sure how we could apply this.

Maybe we could have an option to run the init script only when starting dev mode  but I'm not sure this will work very well in practice as any changes to your model will make your schema outdated.

/cc @yrodiere

**Comment by yrodiere** [comment]
So first, there are certainly users out that who *want* init scripts to be re-executed, because the script drops the whole content of the database before re-creating it, and they want this behavior on app restart.

So, this would need to be configurable, and it's debatable whether it should be the default (I'd say no, but... who knows).

> `testcontainers.reuse.enable=true` is totally outside of our control so I'm not sure how we could apply this.

There is a protected method in `GenericContainer` that we can override to know if the container was reused or not:

```
    protected void containerIsStarted(InspectContainerResponse containerInfo, boolean reused) {
```

If other `*Container` classes expose that, there's a way. That's a big "if" though.

----

On a related note, test containers are only reused if `testcontainers.reuse.enable=true` **and** if we call `.withReuse(true)` on the `*Container` instance.

If container reuse works right now, I suppose we must be calling `.withReuse(true)` indiscriminately, which is agruably a bad idea since users might be setting `testcontainers.reuse.enable=true` because they want reuse for one application, but not for another.

Maybe we should just make "reuse" a devservice option, and make it opt-in? That way we could possibly detect reuse and skip init scripts. We could even offer more than two options: `reuse=false`, `reuse=as-is` (no init scripts whatsoever), `reuse=reinit` (reuse the container, but execute init scripts).

**Comment by yrodiere** [comment]
> There is a protected method in `GenericContainer` that we can override to know if the container was reused or not:
> 
> ```
>     protected void containerIsStarted(InspectContainerResponse containerInfo, boolean reused) {
> ```

FWIW there's this code in `JdbcDatabaseContainer` (the base class for `PostgreSQLContainer`):

```
    protected void containerIsStarted(InspectContainerResponse containerInfo) {
        this.logger().info("Container is started (JDBC URL: {})", this.getJdbcUrl());
        this.runInitScriptIfRequired();
    }
```

So... we should *already* skip init scripts. There's something more to this issue.

**Comment by yrodiere** [comment]
> > There is a protected method in `GenericContainer` that we can override to know if the container was reused or not:
> > ```
> >     protected void containerIsStarted(InspectContainerResponse containerInfo, boolean reused) {
> > ```
> 
> FWIW there's this code in `JdbcDatabaseContainer` (the base class for `PostgreSQLContainer`):
> 
> ```
>     protected void containerIsStarted(InspectContainerResponse containerInfo) {
>         this.logger().info("Container is started (JDBC URL: {})", this.getJdbcUrl());
>         this.runInitScriptIfRequired();
>     }
> ```
> 
> So... we should _already_ skip init scripts. There's something more to this issue.

:facepalm: It's not the same method, the one implemented in `PostgreSQLContainer` doesn't have a `reused` parameter. That explains the behavior, I guess.

On a related note, I created this to document the reuse feature: #37884 . Doesn't fix this issue though.

**Comment by jnbdz** [comment]
I dug into this and can confirm @yrodiere's diagnosis, with the exact mechanism.

**Root cause.** The init scripts from `quarkus.datasource.devservices.init-script-path` are applied with Testcontainers' `withInitScripts(...)`, which are run by `JdbcDatabaseContainer.containerIsStarted(InspectContainerResponse)` → `runInitScriptIfRequired()`. That method override does **not** receive the `reused` flag. `GenericContainer` does have a reuse-aware `containerIsStarted(InspectContainerResponse, boolean reused)`, but its default just delegates to the one-arg version, so a reused container is treated exactly like a fresh one and the scripts run again. For a non-idempotent script (`create user`, `create table`, …) the second start fails.

I verified this with a standalone Testcontainers test (Podman, `TESTCONTAINERS_REUSE_ENABLE=true`): a stock `PostgreSQLContainer` with `withReuse(true).withInitScript(...)` throws on the second start (duplicate object), while a subclass that overrides the two-arg `containerIsStarted` to skip init on reuse starts cleanly.

**Proposed fix** (pushed as a branch, not yet a PR: `SiteNetSoft/quarkus:devservices-skip-init-script-on-reuse`): override the reuse-aware `containerIsStarted(InspectContainerResponse, boolean reused)` in the Quarkus container subclasses and skip the init scripts when `reused` is true, in PostgreSQL/MySQL/MariaDB/MSSQL/DB2/Oracle. Oracle's *privileged* scripts are mounted into the container entrypoint dir and already only run on first boot, so they are unaffected.

**The one open question is the default**, which is what stalled this in 2023. Since Dev Service reuse is opt-in (`quarkus.datasource.devservices.reuse=true` on top of the global `testcontainers.reuse.enable`), I think skip-on-reuse is the right default: opting into reuse means "keep this container and its data", so re-running init that assumes a fresh DB is wrong; users who want drop-and-recreate simply don't enable reuse. That does close the door on @yrodiere's earlier `reinit` idea (reuse a warm container but still re-run init to reset the schema). If you'd rather keep that option, I'm happy to make it a `quarkus.datasource.devservices.init-script-on-reuse` (or a `reuse=false|keep|reinit` mode) instead of an unconditional skip.

If the unconditional skip-on-reuse default is acceptable I'll turn the branch into a PR with a test; if you prefer the configurable variant, only the gating changes and I'll do that instead. cc @yrodiere @gsmet @geoand

**Comment by yrodiere** [comment]
Hey. Please just open the PR, that'll give us a place to discuss this :)

**Comment by geoand** [comment]
I agree with @yrodiere.

I'll also add that the long (AI-generated) summaries are very hard to follow, whereas the code (should, in theory at least) speaks for itself :)
